### ISSUE TITLE
0.12 cty panic: value has no attribute of that name During Import

### ISSUE THREAD

**Issue by bflad** [hashicorp/terraform#20186]
### Terraform Version

```
github.com/hashicorp/terraform v0.12.0-alpha4.0.20190201001850-411dab1dea1a
github.com/zclconf/go-cty v0.0.0-20190130221141-d7fe3fa0020f
```

### Terraform Configuration Files

```hcl
resource "aws_vpc" "example" {
  cidr_block = "10.1.0.0/16"
}

resource "aws_security_group" "example" {
  name = "example"
  vpc_id = "${aws_vpc.example.id}"

  ingress {
    protocol = "6"
    from_port = 80
    to_port = 8000
    cidr_blocks = ["10.0.0.0/8"]
  }

  tags = {
    Name = "example"
  }
}
```

### Debug Output

OSS TeamCity Link: https://HOSTNAME/viewLog.html?buildId=43313&buildTypeId=Aws_ProviderAws&tab=artifacts

### Crash Output

```
=== CONT  TestAccAWSSecurityGroup_importBasic

------- Stderr: -------
panic: value has no attribute of that name

goroutine 984 [running]:
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/zclconf/go-cty/cty.Value.GetAttr(0x4b64760, 0xc0002b33f8, 0x3b4ab20, 0xc000818510, 0xc001b39294, 0xb, 0x0, 0x1, 0xc0003a9860, 0xc0003a9b30)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/zclconf/go-cty/cty/value_ops.go:586 +0x296
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/zclconf/go-cty/cty/msgpack.marshal(0x4b64760, 0xc0002b33f8, 0x3b4ab20, 0xc000818510, 0x4b64760, 0xc0002b3410, 0x0, 0x0, 0x0, 0xc00196a8d0, ...)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/zclconf/go-cty/cty/msgpack/marshal.go:176 +0x488
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/zclconf/go-cty/cty/msgpack.Marshal(0x4b64760, 0xc0002b33f8, 0x3b4ab20, 0xc000818510, 0x4b64760, 0xc0002b3410, 0x0, 0x4d, 0xc000709d60, 0x17, ...)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/zclconf/go-cty/cty/msgpack/marshal.go:35 +0x21c
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin.(*GRPCProvider).ReadResource(0xc000892080, 0xc000709d60, 0x17, 0x4b64760, 0xc0002b33f8, 0x3b4ab20, 0xc000818510, 0x0, 0x0, 0x0, ...)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin/grpc_provider.go:320 +0x10a
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform.(*EvalRefresh).Eval(0xc000e5ae10, 0x4b8b340, 0xc000c3d1e0, 0x2, 0x2, 0x0, 0x0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform/eval_refresh.go:61 +0x40e
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform.EvalRaw(0x4b52280, 0xc000e5ae10, 0x4b8b340, 0xc000c3d1e0, 0x0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform/eval.go:57 +0x167
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc0008f8f00, 0x4b8b340, 0xc000c3d1e0, 0x2, 0x2, 0x7ff818a2d568, 0x67)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform/eval_sequence.go:20 +0xfd
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform.EvalRaw(0x4b522e0, 0xc0008f8f00, 0x4b8b340, 0xc000c3d1e0, 0x3b40560, 0x83c36a7, 0x3784940, 0xc002227de0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform/eval.go:57 +0x167
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform.Eval(0x4b522e0, 0xc0008f8f00, 0x4b8b340, 0xc000c3d1e0, 0xc0008f8f00, 0x4b522e0, 0xc0008f8f00, 0x0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform/eval.go:35 +0x4d
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x3d84a60, 0xc0023a05b0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/terraform/graph.go:90 +0x10d1
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000f85180, 0x3d84a60, 0xc0023a05b0, 0xc000d16380)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/dag/walk.go:392 +0x373
created by github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/dag.(*Walker).Update
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/dag/walk.go:314 +0x9d2
```

### Expected Behavior

No crash. 😄 

### Actual Behavior

Crash. 😅 

### Steps to Reproduce

From an AWS provider branch that has 0.12 update, e.g. `v0.12-sdk-updates-411dab1dea1a`

```
$ TF_ACC=1 go test ./aws -v -timeout 120m -parallel 20 -run=TestAccAWSSecurityGroup_importBasic
```

### Additional Context

Its either related to import or is a strange coincidence to be seemingly only happen with import tests. This also seems to happen with `TestAccAwsDxGateway_importComplex`, so it may be related to or a continuation of the `ResourceData.SetType()` work done recently in #20148

```
=== CONT  TestAccAwsDxGateway_importComplex

------- Stderr: -------
panic: value has no attribute of that name

goroutine 2608 [running]:
github.com/terraform-providers/terraform-provider-aws/vendor/github.com/zclconf/go-cty/cty.Value.GetAttr(0x4b64760, 0xc000cf8aa0, 0x3b4ab20, 0xc001e5ede0, 0xc000e180e0, 0xd, 0x0, 0x1, 0x8, 0x0)
	/opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/zclconf/go-cty/cty/value_ops.go:586 +0x296
```

### References

* Related/continuation of: #20148

**Comment by apparentlymart** [comment]
It looks like there are two different things going on here:

* The `cty` msgpack decoder is crashing in a situation where it should be returning an error.
* Once we fix that, these tests are likely to start returning that new error, so we'll still need to puzzle out why the msgpack value is not valid, if indeed it isn't.

I'm going to work on the `cty` part of this first and then that'll hopefully give the context to understand what, if anything, is going wrong on the Terraform/SDK side.

**Comment by apparentlymart** [comment]
Okay... after some more digging I have a better sense of what is going on here, and indeed it seems that we have two bugs to fix.

During serialization here Terraform Core is passing cty what seems to be an `aws_security_group_rule` object but trying to conform it to the schema for `aws_security_group`. This suggests that #20148 was not a complete enough fix and we're still using the initial resource type to select a schema, rather than the import's overridden resource type. (This makes sense because importing an `aws_security_group` also imports an `aws_security_group_rule` for each of the rules associated with it.)

However, the fact that this managed to cause a panic suggests that something is malfunctioning in `cty`'s type conformance checker, since the msgpack encoder _is_ verifying that the given value conforms to the given type before trying to encode it, and yet that check isn't finding the inconsistency.

As before then, I'm going to first debug the `cty` problem, and then once I have that fixed I'll try to track down where in Terraform Core we're still using the wrong schema. I expect to find it in the area where we try to run a refresh after an import completes, and it'll be using the schema statically selected at graph building time rather than looking up the schema of the resource type indicated in the import response.

**Comment by apparentlymart** [comment]
The upstream `cty` panic is fixed in [`cty` commit `4ca19710f0562cab70f0b3c9cbff0ecc70ee06d1`](https://github.com/zclconf/go-cty/commit/4ca19710f0562cab70f0b3c9cbff0ecc70ee06d1).

I am now working on the second part of the problem that occurs in the Terraform layer, ensuring that we do the follow-up refresh using the right schema for each imported object.

**Comment by apparentlymart** [comment]
The problem in Terraform land was as I expected: the schema was being selected incorrectly. I was incorrect about _where_ that was happening though: the wire protocol marshaling and unmarshaling codepaths were selecting the wrong schema when encoding and decoding the import result, which then caused the rest of Terraform to see a value that didn't actually conform to the schema it claimed to.

Since Terraform's internals tend to assume decoded values are of the expected type (it's the responsibility of the marshal/unmarshal steps to guarantee that), this was producing pretty confusing errors even after fixing the `cty` panic.

There is a fix in #20204.
